### PR TITLE
o/snapstate: allow auto-refresh limited to snaps affected by a specific gating snap

### PR DESCRIPTION
**Pull request by stolowski**
This is to allow calling `snapctl refresh --proceed` from snaps (outside of hooks). The general idea behind the change is:

- we want to trigger auto-refresh but limit updates to snaps only affecting the snap that called --proceed.
- however, we may need to consult other snaps affected by the updates that the requested snap was affected by (meaning running their gate-auto-refresh hooks)
- we will run  gate-auto-refresh hook of the snap that triggered the operation anyway for simplicity/consistency (assuming it will be consistent and say --proceed again, but it's not required of course).
- as a result, some or all refreshes may still be held.

TODO: reorder gate-auto-refresh hooks so that the hook of snap requesting --proceed is run last.

### CODE REVIEW

**Comment by pedronis** [comment]
> TODO: reorder gate-auto-refresh hooks so that the hook of snap requesting --proceed is run first.

no, the agreement we had was to try to run it last

**Comment by stolowski** [comment]
> > TODO: reorder gate-auto-refresh hooks so that the hook of snap requesting --proceed is run first.
> 
> no, the agreement we had was to try to run it last, because it means we know whether we are going to proceed or not and might consider to communicate this

Thanks, I misremembered, updated.

**Review by mardy — APPROVED** [review]
LGTM, but you might want to wait for someone else to review it too :-)

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -2000,6 +2003,36 @@ func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state
 		}
 	}
 
+	// if specific gating snap was given, drop other affected snaps unless
+	// they are affected by same updates as forGatingSnap.
+	if forGatingSnap != "" {
+		if _, ok := affectedSnaps[forGatingSnap]; !ok {
+			// this can happen if after refreshing candidates the snap is no
+			// longer affected by new candidates.
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:2011]: this is a bit confusing to me. Being affected doesn't really change whether there's an update for it or not. We should run an update for it even if nothing is affecting it. The reality is that usually the snap will be added to affectedSnaps because of its own hook but again not sure we should base our decision to do something here on that. The easiest condition to understand is whether forGatingSnap is in `updated` or not at all.

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -2000,6 +2003,36 @@ func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state
...
+		}
+		// filter out unrelated affectedSnaps
+		for gatingSnap, affectedInfo := range affectedSnaps {
+			if gatingSnap == forGatingSnap {
+				continue
+			}
+			var found bool
+			for affectingSnap := range affectedInfo.AffectingSnaps {
+				// does this affectingSnap also affect forGatingSnap?
+				// if so, we will keep it around, meaning we will run its
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:2022]: iiuc what we keep around is the full gatinSnap (that's what the delete below is about), so this comment is a bit confusing and hard to parse.  The condition is really that we keep gating/affectedSnap around if the intersection between its affecting snaps and the one for forGatingSnap is not empty, that means we should call its hook because we can't update the affecting snaps if they affect something else. Which I think is what the code here does correctly, but the comment doesn't explain quite clearly.

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -2000,6 +2003,36 @@ func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state
...
+				// gate-auto-refresh hook because it may potentially hold updates
+				// even if forGatingSnap's hook says "proceed".
+				if _, ok := affectedSnaps[forGatingSnap].AffectingSnaps[affectingSnap]; ok {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:2025]: we don't need the ok bit because this is a map of bools?  also  affectedSnaps[forGatingSnap].AffectingSnaps is fixed for both the loops so we could read it earlier outside? they are the affecting snaps for the forGatingSnap which basically is the snap we will try to update here?

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -2016,10 +2049,20 @@ func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state
...
-	for i, up := range updates {
-		names[i] = up.InstanceName()
+	for _, up := range updates {
+		// if specific gating snap was requested, filter out updates.
+		if forGatingSnap != "" {
+			affecting, ok := affectedSnaps[forGatingSnap]
+			if !ok {
+				continue
+			}
+			if _, ok := affecting.AffectingSnaps[up.InstanceName()]; !ok {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:2061]: it's bool and again it seems we coud read this much earlier,  we are basically finding the intersection between updated & forGatingSnap affecting snaps, also again we should ignore affectingSnaps if the update is actually forGatingSnap

File: overlord/snapstate/snapstate.go [hunk 5]
```diff
@@ -1917,19 +1917,21 @@ func AutoRefresh(ctx context.Context, st *state.State) ([]string, []*state.TaskS
...
 // The state needs to be locked by the caller.
-func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state.TaskSet, error) {
+func autoRefreshPhase1(ctx context.Context, st *state.State, forGatingSnap string) ([]string, []*state.TaskSet, error) {
 	user, err := userFromUserID(st, 0)
 	if err != nil {
 		return nil, nil, err
 	}
 
 	refreshOpts := &store.RefreshOptions{IsAutoRefresh: true}
+	// XXX: should we skip refreshCandidates if forGatingSnap isn't empty (meaning we're handling proceed from a snap)?
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1934]: yes, this is not great because it means we trigger a store op

**Review by pedronis — CHANGES_REQUESTED** [review]
did an incomplete pass, I think the code and comments can be clarified/simplified a bit, also the early return seems attached to a confusing condition.

File: overlord/snapstate/snapstate.go (hunk 2 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go]: Yes, your understanding is correct. I've reworded this comment, hope it makes more sense. Thanks

File: overlord/snapstate/snapstate.go (hunk 3 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go]: Indeed, simplified.

File: overlord/snapstate/snapstate.go (hunk 5 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go:1934]: Hmm, so avoiding this turns out to require some potentially annoying changes, e.g. affectedByRefresh() needs snap.Infos (that comes from candidates/updates) as it looks into info.Hooks. It should be possible but maybe as a followup?

File: overlord/snapstate/snapstate.go (hunk 1 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go]: Fair point, tweaked.

File: overlord/snapstate/snapstate.go (hunk 5 above)
Comment (pedronis) [reply overlord/snapstate/snapstate.go:1934]: maybe it doesn't matter because we will do store operations anyway, also we do need to refresh assertions to do the right thing with validations. It's another reason pointing to the fact that maybe calling proceed outside of the normal refresh hook should be a privileged operation.

File: overlord/snapstate/snapstate.go [hunk 6]
```diff
@@ -1968,6 +1971,20 @@ func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state
...
+	if forGatingSnap != "" {
+		var gatingSnapHasUpdate bool
+		for _, up := range updates {
+			if up.InstanceName() == forGatingSnap {
+				gatingSnapHasUpdate = true
+				break
+			}
+		}
+		if !gatingSnapHasUpdate {
+			// XXX: should this error out instead?
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1983]: this seems fine, as we anyway error here https://github.com/snapcore/snapd/pull/10515/files#diff-61e060364d4189cc8f03d8118c72257b2ce915a1759a83a6b5961340b758507bR660 which is checking the more concrete condition

File: overlord/snapstate/snapstate.go [hunk 7]
```diff
@@ -2016,10 +2061,16 @@ func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state
 
 	// return all names as potentially getting updated even though some may be
 	// held.
-	names := make([]string, len(updates))
+	var names []string
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:2064]: I suppose we could have kept/used:

names := make([]string, 0, len(updates))

instead, that's still covers the common case

File: overlord/snapstate/snapstate.go [hunk 8]
```diff
@@ -2000,6 +2017,34 @@ func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state
...
+		// forGatingSnap and other gating snaps. If so, we need to run
+		// their gate-auto-refresh hooks as well.
+		for gatingSnap, affectedInfo := range affectedSnaps {
+			if gatingSnap == forGatingSnap {
+				continue
+			}
+			var found bool
+			for affectingSnap := range affectedInfo.AffectingSnaps {
+				if snapsAffectingGatingSnap[affectingSnap] {
+					found = true
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:2038]: we never reach here in the tests afaict

File: overlord/snapstate/autorefresh_gating.go [hunk 9]
```diff
@@ -631,3 +632,55 @@ var snapsToRefresh = func(gatingTask *state.Task) ([]*refreshCandidate, error) {
...
+			hasHeld = true
+			break
+		}
+	}
+	if !hasHeld {
+		return fmt.Errorf("no snaps are held by snap %q", gatingSnap)
+	}
+
+	// NOTE: this will unlock and re-lock state for network ops
+	// XXX: should we refresh assertions (just call AutoRefresh()?)
```
Comment (pedronis) [inline overlord/snapstate/autorefresh_gating.go:664]: yes, we should do something about refreshing assertions, mainly for the usual reasons which is validations. We might want to do that separately though because I think the work on validation-sets will also need to tweak how we organize refreshing assertions

**Review by pedronis — COMMENTED** [review]
thanks for all the changes, the code seems a bit simpler/clearer now. I answered some questions. It seems we are missing tests about the intersection case though, unless I'm confused

File: overlord/snapstate/snapstate.go (hunk 8 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go:2037]: Indeed, added.

File: overlord/snapstate/snapstate.go (hunk 7 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go]: Done.

**Review by pedronis — APPROVED** [review]
thanks, this one probably needs a 3rd review

**Review by anonymouse64 — APPROVED** [review]
one small typo, otherwise lgtm

**Comment by anonymouse64** [comment]
err sorry there wasn't a typo, that was a typo on my part /o\
